experimental/Schemes/WeilDivisor.jl

```diff
@@ -660,15 +705,83 @@ end
   return coefficient_dict(D)[C] == 1
 end
 
-is_irreducible(D::WeilDivisor) = is_prime(D)
+is_irreducible(D::AbsWeilDivisor) = is_prime(D)
 
 function order_on_divisor(
     f::VarietyFunctionFieldElem,
-    D::WeilDivisor;
+    D::AbsWeilDivisor;
     check::Bool=true
   )
   @check is_prime(D) || error("divisor must be prime")
   I = components(D)[1]
   return order_on_divisor(f, I, check=false)
 end
 
+########################################################################
+# Implement the interface for AbsAlgebraicCycle and WeilDivisor for    
+# ToricDivisor
+#
+# Because of the inclusion order of the files, this has to be postponed
+# until here.
+########################################################################
+
+scheme(td::ToricDivisor) = toric_variety(td)
+
+### Construct the Weil divisor associated to `td`.
+@attr WeilDivisor function underlying_divisor(td::ToricDivisor)
+  X = scheme(td)
+  generating_divisors = _torus_invariant_weil_divisors(X)
+  return sum(a*D for (a, D) in zip(coefficients(td), generating_divisors))
+end
+
+# Compute the ideal sheaves and Weil divisors of the rays of `fan(X)` 
+# according to the "orbit-cone-correspondence", Theorem 3.2.6 in 
+# Cox-Little-Schenk.
+#
+# We would like to use Equation (3.2.7) from page 121 for this, 
+# but that equation is wrong (it always produces the unit ideal). 
+# We correct it as follows. Let τ ∈ Σ be a ray and σ ∈ Σ a maximal 
+# cone. The ideal of the intersection of the orbit with the chart 
+#   I(O(τ) ∩ Uσ) ⊂ ℂ[σ ̌∩ M] 
+# is either 
+#   1) the unit ideal if τ ⊄ σ
+#   2) the ideal generated by the elements in 
+#         σ ̌∩ M ∖ τ⟂ 
+#      i.e. the elements of σ ̌∩ M which are not in the annihilator 
+#      of τ. 
+# To get hold of these generators, we make the following assumption
+#
+#   (*) When {v₁,…,vᵣ} is a Hilbert basis of σ ̌∩ M, then in case 2) 
+#       above the elements in σ ̌∩ M ∖ τ⟂  are precisely those in 
+#         Σ vⱼ + σ ̌∩ M 
+#       for those vⱼ which are not in τ⟂. Hence, we can use the 
+#       monomials corresponding to those vⱼ as generators of the 
+#       ideal.
+#
+# Edit: It turned out to be a misprint in the online version of the 
+# CLS book. In the printed version, which might not be accessible to 
+# all of us, the formula says exactly what we say in 2). 
+@attr Vector{<:WeilDivisor} function _torus_invariant_weil_divisors(X::NormalToricVariety)
+  ray_list = rays(fan(X))
+  ideal_sheaves = Vector{IdealSheaf}()
+  for tau in ray_list
+    tau_dual = polarize(cone(tau))
+    ideal_dict = IdDict{AbsSpec, Ideal}()
+    for U in affine_charts(X)
+      if !(tau in cone(U))
+        ideal_dict[U] = ideal(OO(U), one(OO(U)))
+        continue
+      end
+      sigma_dual = weight_cone(U)
+      hb = hilbert_basis(sigma_dual)
+      selection = [v for v in hb if !(-v in tau_dual)]
+      x = gens(OO(U))
+      ideal_dict[U] = ideal(OO(U), [x[i] for i in 1:length(x) if !(-hb[i] in tau_dual)])
```

Comment:
Can selection be removed? Or alternatively, can it be used in creating `ideal_dict[U]`?